pypicloud/cache/dynamo.py

```diff
@@ -22,6 +24,82 @@
 LOG = logging.getLogger(__name__)
 
 
+def _clear_summary(summary):
+
+    """ Clear out the field of a PackageSummary object """
+    summary.stable = None
+    summary.unstable = '0'
+    summary.last_modified = datetime.fromtimestamp(0).replace(tzinfo=UTC)
+    return summary
+
+
+def calculate_cache_updates(cache_packages, cache_summaries, storage_packages, summary_factory):
+
+    """ Calculates what changes need to happen in the cache to update it. """
+    # Clear out non-key fields of the cached summaries to purge stable and
+    # unstable fields.
+    cached_summ_by_name = dict(
+        (summ.name, _clear_summary(summ),)
+        for summ in cache_summaries
+    )
+    cached_pkg_by_filename = dict(
+        (pkg.filename, pkg,)
+        for pkg in cache_packages
+    )
+
+    new_packages = set()
+    seen_packages = set()
+    new_summaries_by_name = dict()
+    seen_summaries = set()
+    for pkg in storage_packages:
+        try:
+            cached_pkg = cached_pkg_by_filename[pkg.filename]
+        except KeyError:
+            new_packages.add(pkg)
+        else:
+            seen_packages.add(cached_pkg)
+
+            # Skip over duplicate files (same filename, different paths).
+            # XXX Is this acceptable? How should I decide this?
+            if cached_pkg.last_modified > pkg.last_modified:
+                continue
+
+            cached_pkg.filename = pkg.filename
+            cached_pkg.name = pkg.name
+            cached_pkg.version = pkg.version
+            cached_pkg.last_modified = pkg.last_modified
+            cached_pkg.data = pkg.data
+
+        try:
+            summ = cached_summ_by_name[pkg.name]
+        except KeyError:
+            summ = new_summaries_by_name.setdefault(
+                pkg.name,
+                summary_factory(pkg),
+            )
+        else:
+            seen_summaries.add(summ)
+
+        summ.update_with(pkg)
+
+    new_summaries = new_summaries_by_name.itervalues()
+    stale_packages = set(cached_pkg_by_filename.itervalues()) - seen_packages
+    stale_summaries = set(cached_summ_by_name.itervalues()) - seen_summaries
+
+    return (
+        (
```

Comment:
Yeah. I was trying to avoid creating a new data type just for one use site. However, this got pretty comical and you also felt the inclination.

Moving for that.